# Web Mining and Applied NLP (44-620)

## Requests, JSON, and NLP

### Student Name: Emily Zimmerman

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [104]:
import requests
import json

ARTIST='Edgar Allan Poe'
SONG = 'A Dream Within A Dream'
URL = f'https://poetrydb.org/author,title/{ARTIST};{SONG}'
result = json.loads(requests.get(URL).text)

with open("lyrics.json", "w") as outfile:
    json.dump(result, outfile)

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [115]:
import requests
import json
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob
import pickle


file = open('lyrics','r')
song = json.load(file)
lyrics = song[0]['lines']
string = ' '.join(lyrics)
print(string)


nlp = spacy.load('en_core_web_sm')
nlp.add_pipe("spacytextblob")
doc = nlp(string)
print(doc._.blob.polarity)

#QUESTION ANSWER:
#I think the lyrics have more positive connotation, the polarity score is above 0 and upon reading the text, the words seem very postitive. 


Take this kiss upon the brow! And, in parting from you now, Thus much let me avow-- You are not wrong, who deem That my days have been a dream: Yet if hope has flown away In a night, or in a day, In a vision or in none, Is it therefore the less _gone_? _All_ that we see or seem Is but a dream within a dream.  I stand amid the roar Of a surf-tormented shore, And I hold within my hand Grains of the golden sand-- How few! yet how they creep Through my fingers to the deep While I weep--while I weep! O God! can I not grasp Them with a tighter clasp? O God! can I not save _One_ from the pitiless wave? Is _all_ that we see or seem But a dream within a dream?
0.055555555555555546


3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [64]:
import requests
import json

print("Enter the artists (author) name:")
ARTIST = input()
print("Enter the song (poem) name:")
SONG = input()
print("Enter the filename")
FILENAME = input()

def poem(ARTIST, SONG, FILENAME):
    URL = f'https://poetrydb.org/author,title/{ARTIST};{SONG}'
    result = json.loads(requests.get(URL).text)
    print(result)
    with open(FILENAME, "w") as outfile:
        json.dump(result, outfile)
    
poem(ARTIST, SONG, FILENAME)

Enter the artists (author) name:


 Artist


Enter the song (poem) name:


 Song


Enter the filename


 Name


{'status': 404, 'reason': 'Not found'}


4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [120]:
import requests
import json
import spacy

print("Enter the file name:")
FILENAME = input()
print("Enter song name:")
song = input()

def polarity(FILENAME):
    file = open(FILENAME,'r')
    song = json.load(file)
    lyrics = song[0]['lines']
    string = ' '.join(lyrics)
    nlp = spacy.load('en_core_web_sm')
    nlp.add_pipe("spacytextblob")
    doc = nlp(string)
    print(doc._.blob.polarity)

print("Polarity of", song)
polarity(FILENAME)

Enter the file name:


 DowsonSong


Enter song name:


 song


Polarity of song
0.22500000000000003


The Shelley song (polarity of -.02) did match my understanding of the lyrics however I was surprised that it was closer to 0 and not closer to -1. 
The Wheatley song (polarity of (0.18) did not at all match my understanding of the lyrics, there were a lot of evil lyrics. 
The Dowson song (polarity of 0.225) was a bit lower than I expected, as the poem seems to be a love poem. 
I think that the connotation of words can affect how the polarity reads them because polarity cannot read context.